src/main/java/com/google/api/generator/gapic/model/HttpRuleBindings.java

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.gapic.model;
+
+import com.google.api.generator.gapic.utils.JavaStyle;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableSet;
+import java.util.Set;
+
+@AutoValue
+public abstract class HttpRuleBindings {
+  // private static final Pattern PATH_PARAM_PATTERN = Pattern.compile("[]")
+  public abstract String httpVerb();
+
+  public abstract String pattern();
+
+  public abstract Set<String> pathParameters();
+
+  public abstract Set<String> queryParameters();
+
+  public abstract Set<String> bodyParameters();
+
+  // Private.
+  public static HttpRuleBindings.Builder builder() {
+    return new AutoValue_HttpRuleBindings.Builder()
+        .setHttpVerb("")
```

Comment:
They don't but that is basically indication of empty `HttpRuleBindings`, which, on its own is a valid thing (represents a method in protobuf service without http annotation at all). So it is basically a choice of empty string vs `null` here. Most best practices prefer empty values instead of nulls (even, for exmaple, protobuf stubs themselves). If you have a strong opinion about it, I can change it to `null`.

Reply:
To clarify, would it be valid for an HTTP annotation to be present, but have an empty string as the pattern, and an empty string as the HTTP verb? More in the next comment.

Reply:
No, it would not be valid. But that class is not strictly speaking a representation of `option (google.api.http)`. It is initialized from that rule, but not necessarily repeats it. Basically the mental model around it is:
- `HttpRuleBindings` is initialized from `google.api.http` option but is not its strict representation
- Basically the main difference is this: 
  - for each `google.api.http` there will be `HttpRuleBindings`
  - but there will be `HttpRuleBindings` even when there is no `google.api.http`
  
Basically the design decision is to use `HttpRuleBindings` as a representation of both situations: `google.api.http` with some values in it and absence of `google.api.http` whatsoever.